File: other_languages/go/redblacktree/insert.go
package redblacktree

// Insert inserts new node into the red black tree
func (tree *RBTree) Insert(data int) {
	var insertedNode *Node
	if tree.Root == nil {
		tree.Root = NewNode(data, RED)
		insertedNode = tree.Root
	} else {
		node := tree.Root
		for true {
			if data < node.Data {
				if node.Left == nil {
					node.Left = NewNode(data, RED)
					insertedNode = node.Left
					break
				} else {
					node = node.Left
				}
			} else if data > node.Data {
				if node.Right == nil {
					node.Right = NewNode(data, RED)
					insertedNode = node.Right
					break
				} else {
					node = node.Right
				}
			} else {
				node.Data = data
				return
			}
		}
		insertedNode.Parent = node
	}
	tree.insertCase1(insertedNode)
	tree.Size++
}

func (tree *RBTree) insertCase1(node *Node) {
	if node.Parent == nil {
		node.Color = BLACK
	} else {
		tree.insertCase2(node)
	}
}

func (tree *RBTree) insertCase2(node *Node) {
	if nodeColor(node.Parent) == BLACK {
		return
	}
	tree.insertCase3(node)
}

func (tree *RBTree) insertCase3(node *Node) {
	uncle := node.uncle()
	if nodeColor(uncle) == RED {
		node.Parent.Color = BLACK
		uncle.Color = BLACK
		node.grandparent().Color = RED
		tree.insertCase1(node.grandparent())
	} else {
		tree.insertCase4(node)
	}
}

func (tree *RBTree) insertCase4(node *Node) {
	grandparent := node.grandparent()
	if node == node.Parent.Right && node.Parent == grandparent.Left {
		tree.leftRotate(node.Parent)
		node = node.Left
	} else if node == node.Parent.Left && node.Parent == grandparent.Right {
		tree.rightRotate(node.Parent)
		node = node.Right
	}
	tree.insertCase5(node)
}

func (tree *RBTree) insertCase5(node *Node) {
	node.Parent.Color = BLACK
	grandparent := node.grandparent()
	grandparent.Color = RED
	if node == node.Parent.Left && node.Parent == grandparent.Left {
		tree.rightRotate(grandparent)
	} else if node == node.Parent.Right && node.Parent == grandparent.Right {
		tree.leftRotate(grandparent)
	}
}

File: RedBlackTree/insert.cpp
#include "RedBlackTree.hpp"

/**
 * inserts new integer data into the tree at the position, so that
 * the red black tree property is maintained.
 * @param data new integer data to be inserted
*/
void RedBlackTree::insert(int data) {
    Node* insertedNode = nullptr;
    if (root == nullptr) {
        root = new Node(data, RED);
        insertedNode = root;
    } else {
        Node* node = root;
        while (true) {
            if (data < node->data) {
                if (node->left == nullptr) {
                    node->left = new Node(data, RED);
                    insertedNode = node->left;
                    break;
                } else {
                    node = node->left;
                }
            } else if (data > node->data) {
                if (node->right == nullptr) {
                    node->right = new Node(data, RED);
                    insertedNode = node->right;
                    break;
                } else {
                    node = node->right;
                }
            } else {
                node->data = data;
                return;
            }
        }
        insertedNode->parent = node;
    }
    insertCase1(insertedNode);
    size++;
}

void RedBlackTree::insertCase1(Node* node) {
    if (node->parent == nullptr) {
        node->color = BLACK;
    } else {
        insertCase2(node);
    }
}

void RedBlackTree::insertCase2(Node* node) {
    if (getNodeColor(node->parent) == BLACK) return;
    insertCase3(node);
}

void RedBlackTree::insertCase3(Node* node) {
    Node* uncleNode = uncle(node);
    if (getNodeColor(uncleNode) == RED) {
        node->parent->color = BLACK;
        uncleNode->color = BLACK;
        grandparent(node)->color = RED;
        insertCase1(grandparent(node));
    } else {
        insertCase4(node);
    }
}

void RedBlackTree::insertCase4(Node* node) {
    Node* grandparentNode = grandparent(node);
    if (node == node->parent->right && node->parent == grandparentNode->left) {
        leftRotate(node->parent);
        node = node->left;
    } else if (node == node->parent->left &&
               node->parent == grandparentNode->right) {
        rightRotate(node->parent);
        node = node->right;
    }
    insertCase5(node);
}

void RedBlackTree::insertCase5(Node* node) {
    node->parent->color = BLACK;
    Node* grandparentNode = grandparent(node);
    grandparentNode->color = RED;
    if (node == node->parent->left && node->parent == grandparentNode->left) {
        rightRotate(grandparentNode);
    } else if (node == node->parent->right &&
               node->parent == grandparentNode->right) {
        leftRotate(grandparentNode);
    }
}